File: src/main/java/gg/mineral/practice/util/messages/Message.java
package gg.mineral.practice.util.messages;

import org.bukkit.entity.Player;

public abstract class Message {
    protected final StringBuilder messageBuilder;

    public Message(String string) {
        this.messageBuilder = new StringBuilder(string);
    }

    public void add(Message message) {
        messageBuilder.append(message.toString());
    }

    public void send(Player player) {
        player.sendMessage(toString());
    }

    @Override
    public String toString() {
        return messageBuilder.toString();
    }

    @Override
    public abstract Message clone();
}

File: src/main/kotlin/gg/mineral/practice/util/messages/Message.kt
package gg.mineral.practice.util.messages

import org.bukkit.command.CommandSender

abstract class Message(string: String) {
    protected val messageBuilder by lazy {
        prepend(StringBuilder(string))
    }

    fun add(message: Message): StringBuilder = messageBuilder.append(message.toString())

    open fun send(sender: CommandSender) = sender.sendMessage(toString())

    abstract fun prepend(builder: StringBuilder): StringBuilder

    final override fun toString() = messageBuilder.toString()

    abstract fun clone(): Message
}
